sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/DataSourceV2Strategy.scala

```diff
@@ -0,0 +1,95 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.Strategy
+import org.apache.spark.sql.catalyst.expressions._
+import org.apache.spark.sql.catalyst.planning.PhysicalOperation
+import org.apache.spark.sql.catalyst.plans.logical.LogicalPlan
+import org.apache.spark.sql.execution.{FilterExec, ProjectExec, SparkPlan}
+import org.apache.spark.sql.execution.datasources.DataSourceStrategy
+import org.apache.spark.sql.sources.Filter
+import org.apache.spark.sql.sources.v2.reader.downward.{CatalystFilterPushDownSupport, ColumnPruningSupport, FilterPushDownSupport}
+
+object DataSourceV2Strategy extends Strategy {
+  // TODO: write path
+  override def apply(plan: LogicalPlan): Seq[SparkPlan] = plan match {
+    case PhysicalOperation(projects, filters, DataSourceV2Relation(output, reader)) =>
+      val attrMap = AttributeMap(output.zip(output))
+
+      val projectSet = AttributeSet(projects.flatMap(_.references))
+      val filterSet = AttributeSet(filters.flatMap(_.references))
+
+      // Match original case of attributes.
+      // TODO: nested fields pruning
+      val requiredColumns = (projectSet ++ filterSet).toSeq.map(attrMap)
+      reader match {
+        case r: ColumnPruningSupport =>
+          r.pruneColumns(requiredColumns.toStructType)
+        case _ =>
+      }
+
+      val stayUpFilters: Seq[Expression] = reader match {
+        case r: CatalystFilterPushDownSupport =>
+          r.pushCatalystFilters(filters.toArray)
+
+        case r: FilterPushDownSupport =>
```

Comment:
Looks like `CatalystFilterPushDownSupport` and `FilterPushDownSupport` are exclusive?

Reply:
yea we can't prevent users to implement them both, and we will pick `CatalystFilterPushDownSupport` over `FilterPushDownSupport`. Let me document it.

Reply:
Considering that there is a translation between Catalyst filters and Filters, it's probably worth _just_ doing the catalyst one, and providing the user with the translator if they want to do the Filter approach?

Reply:
can FilterPushDownSupport be an interface which extends CatalystFilterPushDownSupport and provides a default impl of pruning the catalyst flter? Like, this code can just go there as a method:

```
interface FilterPushDownSupport extends CatalystFilterPushDownSupport {
    List<Filter> pushFilters(List<Filter> filters);

    default List<Expression> pushCatalystFilters(List<Expression> filters) {
        Map<Filter, Expression> translatedMap = new HashMap<>();
        List<Filter> nonconvertiblePredicates = new ArrayList<>();

        for (Expression catalystFilter : filters) {
            Optional<Filter> translatedFilter = DataSourceStrategy.translateFilter(catalystFilter);
            if (translatedFilter.isPresent()) {
                translatedMap.put(translatedFilter.get(), catalystFilter);
            } else {
                nonconvertiblePredicates.add(catalystFilter);
            }
        }

        List<Filter> unhandledFilters = pushFilters(new ArrayList<>(translatedMap.values()));
        return Stream.concat(
            nonconvertiblePredicates.stream(),
            unhandledFilters().stream().map(translatedMap::get))
           .collect(toList());
    }
}
```

and we can trivially ignore the interface confusion (it's truly confusing if you can implement two interfaces)

Reply:
like, we might as well not document it if the code can document it

Reply:
good idea!

Reply:
By doing so, do we still need to match both `CatalystFilterPushDownSupport` and `FilterPushDownSupport` here?

Reply:
After some attempts, I went back with 2 individual interfaces. The reason is that, a) `CatalystFilterPushDownSupport` is an unstable interface, and it looks weird to let a stable interface extend an unstable one. b) the logic that converts expressions to public filters belongs to Spark internal, and we may change it in the future, so we should not put these codes in a public interface. We may have a risk of breaking compatibility for this interface.